internal/cmd/auth/login/login.go

```
@@ -0,0 +1,95 @@
+package login
+
+import (
+	"errors"
+	"fmt"
+	"net/url"
+	"os"
+
+	"github.com/spf13/cobra"
+
+	"code-intelligence.com/cifuzz/internal/access_tokens"
+	"code-intelligence.com/cifuzz/internal/api"
+	"code-intelligence.com/cifuzz/internal/cmdutils"
+	"code-intelligence.com/cifuzz/pkg/dialog"
+	"code-intelligence.com/cifuzz/pkg/log"
+)
+
+func New() *cobra.Command {
+	var bindFlags func()
+
+	cmd := &cobra.Command{
+		Use:     "login",
+		Short:   "Login to a remote backend",
+		Long:    `This command is used to authenticate with a remote backend to manage findings.`,
+		Example: "$ cifuzz auth login",
+		PreRunE: func(cmd *cobra.Command, args []string) error {
+			bindFlags()
+			return nil
+		},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			server := cmd.PersistentFlags().Lookup("server").Value.String()
+
+			// check if server URL is valid, if not try to add https://
+			_, err := url.ParseRequestURI(server)
+			if err != nil {
+				server = fmt.Sprintf("https://%s", server)
+			}
+
+			return Login(server)
+		},
+	}
+	bindFlags = cmdutils.AddFlags(cmd,
+		cmdutils.AddServerFlag,
+	)
+	return cmd
+}
+
+func isTokenValid(server string, token string) bool {
+	client := api.Client{Server: server}
+	// TOOD: Change this to use another check without querying projects
+	_, err := client.ListProjects(token)
+	return err == nil
+}
+
+func Login(server string) error {
+	// Obtain the API access token
+	token := os.Getenv("CIFUZZ_API_TOKEN")
+	if token == "" {
+		token = access_tokens.Get(server)
+	}
+
+	if token == "" {
+		msg := fmt.Sprintf(`Enter an API access token and press Enter. You can generate a token for
+your account at %s/dashboard/settings/account.`+"\n", server)
+		token, err := dialog.ReadSecret(msg, os.Stdin)
+		if err != nil {
+			return err
+		}
+
+		// Try to authenticate with the access token
+		if !isTokenValid(server, token) {
+			return errors.New("failed to authenticate with the provided API access token")
```

This will cause hard to debug errors if we don't print the error which the `ListProjects` call failed with.

---

I changed the `isTokenValid` to print the error when it is not nil. Is this ok?
We should update the function nonetheless once another API call for testing auth with the backend is available. Listing Projects works, but it's not ideal..

---

I now that `ListProjects` already handles request errors by printing the error and returning a silent error. That was fine when it was an internal function of the `remote_run` package, but now that it's an exported function of the `api` package, it should leave it to the caller to handle errors. 

Ideally, we return a custom error which allows the caller to inspect the status code, which allows us to handle actual authentication errors differently from other errors in the `isTokenValid` function.

---

Done